cpp/src/arrow/compute/exec/key_compare.cc

```diff
@@ -17,250 +17,398 @@
-          rows_left, rows_right);
-    } else {
-      num_rows_already_processed =
-          CompareFixedLength_avx2(num_rows_to_compare, left_to_right_map,
-                                  match_bytevector, fixed_length, rows_left, rows_right);
-    }
+  if (ctx->has_avx2()) {
+    num_processed = CompareBinaryColumnToRow_avx2(
+        use_selection, offset_within_row, num_rows_to_compare, sel_left_maybe_null,
+        left_to_right_map, ctx, col, rows, match_bytevector);
   }
 #endif
 
-  typedef void (*CompareFixedLengthImp_t)(uint32_t, uint32_t, const uint16_t*,
-                                          const uint32_t*, uint8_t*, uint32_t,
-                                          const uint8_t*, const uint8_t*);
-  static const CompareFixedLengthImp_t CompareFixedLengthImp_fn[] = {
-      CompareFixedLengthImp<false, 1>, CompareFixedLengthImp<false, 2>,
-      CompareFixedLengthImp<false, 0>, CompareFixedLengthImp<true, 1>,
-      CompareFixedLengthImp<true, 2>,  CompareFixedLengthImp<true, 0>};
-  int dispatch_const = (use_selection ? 3 : 0) +
-                       ((fixed_length <= 8) ? 0 : ((fixed_length <= 16) ? 1 : 2));
-  CompareFixedLengthImp_fn[dispatch_const](
-      num_rows_already_processed, num_rows_to_compare, sel_left_maybe_null,
-      left_to_right_map, match_bytevector, fixed_length, rows_left, rows_right);
-}
+  uint32_t col_width = col.metadata().fixed_length;
+  if (col_width == 0) {
+    int bit_offset = col.bit_offset(1);
+    CompareBinaryColumnToRowHelper<use_selection>(
+        offset_within_row, num_processed, num_rows_to_compare, sel_left_maybe_null,
+        left_to_right_map, ctx, col, rows, match_bytevector,
+        [bit_offset](const uint8_t* left_base, const uint8_t* right_base,
+                     uint32_t irow_left, uint32_t offset_right) {
+          uint8_t left = BitUtil::GetBit(left_base, irow_left + bit_offset) ? 0xff : 0x00;
+          uint8_t right = right_base[offset_right];
+          return left == right ? 0xff : 0;
+        });
+  } else if (col_width == 1) {
+    CompareBinaryColumnToRowHelper<use_selection>(
+        offset_within_row, num_processed, num_rows_to_compare, sel_left_maybe_null,
+        left_to_right_map, ctx, col, rows, match_bytevector,
+        [](const uint8_t* left_base, const uint8_t* right_base, uint32_t irow_left,
+           uint32_t offset_right) {
+          uint8_t left = left_base[irow_left];
+          uint8_t right = right_base[offset_right];
+          return left == right ? 0xff : 0;
+        });
+  } else if (col_width == 2) {
+    CompareBinaryColumnToRowHelper<use_selection>(
+        offset_within_row, num_processed, num_rows_to_compare, sel_left_maybe_null,
+        left_to_right_map, ctx, col, rows, match_bytevector,
+        [](const uint8_t* left_base, const uint8_t* right_base, uint32_t irow_left,
+           uint32_t offset_right) {
+          uint16_t left = reinterpret_cast<const uint16_t*>(left_base)[irow_left];
```

Comment:
Are these guaranteed to be aligned? If so, please add an assertion and a comment.

Reply:
I added alignment checks when guaranteed and SafeLoads when not.